### DISCUSSION TITLE
Illegal instruction (SIGILL) (core dumped) when importing netket

### DISCUSSION THREAD

**Discussion by JanaMensikova** [netket/netket#670] [Q&A] [answered]
Hello,
I have a problem executing my python programs when importing netket.
I installed netket according to the instuctions on netket website and it went succesfully, but when I try to run a program where command import netket is included, the terminal writes this: Illegal instruction (SIGILL) (core dumped).
I have netket version 3.0b1.post8.

I use Linux Mint distribution with following parameters:

System:
Kernel: 5.4.0-58-generic x86_64 bits: 64
compiler: gcc v: 9.3.0 Desktop: Cinnamon 4.8.5
Distro: Linux Mint 20.1 Ulyssa base: Ubuntu 20.04 focal

Machine:
Type: Laptop System: HP product: HP Laptop 17-by0xxx
v: Type1ProductConfigId serial: <superuser/root required>
Mobo: HP model: 84CE v: KBC Version 89.30
serial: <superuser/root required> UEFI: Insyde v: F.53 date: 12/13/2019

CPU:
Topology: Quad Core model: Intel Pentium Silver N5000 bits: 64 type: MCP
arch: Goldmont Plus rev: 1 L2 cache: 4096 KiB
 
Graphics:
Device-1: Intel UHD Graphics 605 vendor: Hewlett-Packard driver: i915
v: kernel bus ID: 00:02.0

I had exactly the same problem with Ubuntu 20.10 (Groovy Gorilla), so I tried Mint distribution instead, but the problem remains.

Thank you for your advice.
Jana Menšíková

**Comment by PhilipVinc** [comment]
CAN you give more details? How did you install netket?
What Python?
Did you also install mpi dependencies or not?
Can you import jax/jaxlib/(mpi4jax) independently?

What is the full stack trace?

**Comment by JanaMensikova** [comment]
I installed netket with command pip install --pre netket (according to the website).
I have python 3.8.5 installed.
I also installed mpi dependencies.
When I import jaxlib, it works, but when I import only jax/mpi4jax, the terminal states: Illegal instruction (SIGILL) (core dumped).

**Comment by PhilipVinc** [comment]
Sorry I don't understand... 
When you import only jax do you get a crash? Or only when you import mpi4jax?

**Comment by JanaMensikova** [comment]
When I import jax, I get a crash.

**Comment by JanaMensikova** [comment]
And the same goes when I import mpi4jax.

**Comment by PhilipVinc** [comment]
Yea well, mpi4jax imports jax internally so if you crash importing jax you will crash importing mpi4jax.

I have no idea what might be wrong. I suggest you open an issue on google/jax repository.

**Comment by gcarleo** [comment]
maybe you can try to install it without conda? conda can be problematic...

**Comment by PhilipVinc** [comment]
Ah wait, maybe I know what it is.
I think recently jaxlib started requiring advanced SIMD instructions...
Can you try downgrading by two minor versions the jaxlib you have installed?

**Comment by JanaMensikova** [comment]
I didn't use conda, I installed everything with pip.

Oh, ok, I will try to downgrade it, thanks.

**Comment by PhilipVinc** [comment]
Try jaxlib 0.1.61 (0.1.62 requires AVX, and I think your processor does not have that)

**Comment by JanaMensikova** [comment]
I've had version 0.1.64, so I uninstalled it and installed 0.1.61 instead.
When I try to run the program, the terminal writes this: ValueError: jaxlib is version 0.1.61, but this version of jax requires version 0.1.64.

**Comment by PhilipVinc** [comment]
Decrease jax version 1 by 1 until you get the one that works

**Comment by PhilipVinc** [comment, marked as answer]
Yes i'm now sure it's that. 
If you read Jax changelog for version [0.1.62](https://jax.readthedocs.io/en/latest/changelog.html#jaxlib-0-1-62-march-9-2021) you see that they enabled AVX by default. however [your cpu](https://ark.intel.com/content/www/us/en/ark/products/128990/intel-pentium-silver-n5000-processor-4m-cache-up-to-2-70-ghz.html) does not have AVX but only SSE4.2 and so you get an Illegal instruction.

I have to say that jax could at least error humanly and say what the problem is instead of just giving cryptic informations...

jax 0.2.11 requires jaxlib 0.1.62 so you probably must go back to jax 0.2.10
Mpi4jax requires jaxlib 0.1.62 too, so you will have to uninstall mpi4jax too (mpi4jax is basically broken for all jaxlib versions between 0.1.58 and 0.1.61, inclusive. See [mpi4jax#46](https://github.com/mpi4jax/mpi4jax/issues/46))

**Comment by PhilipVinc** [comment]
Alternatively, you can build jaxlib for yourself disabling AVX. You must pass the flags specified in the changelog of version 0.1.62 i linked above.
I don't think it will be super-easy, though.

(By the way, sorry if my first messages sounded a bit cold (they weren't supposed to be) but I was trying to answer quickly from my phone, and I can be a bit direct sometimes.)

**Comment by JanaMensikova** [comment]
Ok, thank you very much, I will try it, hope it will work =)

Don't worry about that, I apperciate your prompt reaction.

**Comment by PhilipVinc** [comment]
Let me know how it goes.
If that was indeed the cause I'll move this issue to Discussions and Change the title to better reflect what the issue was.

That way, if anybody stumbles on that again at least he'll find some decent documentation on it, since jax doesn't provide it.

**Comment by JanaMensikova** [comment]
Ok, now I have jax 0.2.10 and jaxlib 0.1.61, when I run the program, the terminal writes:

2021-04-29 16:11:46.739706: W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-04-29 16:11:47.005848: W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
Traceback (most recent call last):
  File "zkouska.py", line 1, in <module>
    import netket
  File "/home/janca/.local/lib/python3.8/site-packages/netket/__init__.py", line 24, in <module>
    from . import utils
  File "/home/janca/.local/lib/python3.8/site-packages/netket/utils/__init__.py", line 36, in <module>
    from .history import History, accum_in_tree, accum_histories_in_tree
  File "/home/janca/.local/lib/python3.8/site-packages/netket/utils/history.py", line 23, in <module>
    from .numbers import dtype, is_scalar
  File "/home/janca/.local/lib/python3.8/site-packages/netket/utils/numbers.py", line 10, in <module>
    ArrayT = Union[np.ndarray, jaxlib.xla_extension.DeviceArray, jax.core.Tracer]
AttributeError: module 'jaxlib.xla_extension' has no attribute 'DeviceArray'

Is there some problem with tensorflow? I have version 2.4.1 installed.

**Comment by PhilipVinc** [comment]
how did you install jaxlib? did you do pip install jaxlib==0.1.61 ? Or did you install it from the url of google to have nvidia gpu support?

It seems that jaxlib is trying to me that jax is trying to load cuda but you have no cuda installed (according to your original post, you don't have a nvidia gpu in your computer). 

The fact that he says tensorflow is that jaxlib is basically a part of tensorflow.

--

About the error, ah. 
It's because in netket too we expect jaxlib 0.1.62 at least. I will try to fix this. It should work fine for the time being however, you might get errors in some cases. Do let me know.

**Comment by JanaMensikova** [comment]
I installed jaxlib with this command: pip install -U jax jaxlib==0.1.61+cuda111 -f https://storage.googleapis.com/jax-releases/jax_releases.html.

**Comment by PhilipVinc** [comment]
Do you actually have a cuda gpu?

**Comment by JanaMensikova** [comment]
It seems I don't have a cuda gpu.

**Comment by PhilipVinc** [comment]
So just install jaxlib with pip install jaxLob==x.y.z 
In any case the warning is meaningless

**Comment by JanaMensikova** [comment]
Ok, so now I have installed jaxlib according to your last advice, I still get this error:
Traceback (most recent call last):
  File "zkouska.py", line 1, in <module>
    import netket
  File "/home/janca/.local/lib/python3.8/site-packages/netket/__init__.py", line 24, in <module>
    from . import utils
  File "/home/janca/.local/lib/python3.8/site-packages/netket/utils/__init__.py", line 36, in <module>
    from .history import History, accum_in_tree, accum_histories_in_tree
  File "/home/janca/.local/lib/python3.8/site-packages/netket/utils/history.py", line 23, in <module>
    from .numbers import dtype, is_scalar
  File "/home/janca/.local/lib/python3.8/site-packages/netket/utils/numbers.py", line 10, in <module>
    ArrayT = Union[np.ndarray, jaxlib.xla_extension.DeviceArray, jax.core.Tracer]
AttributeError: module 'jaxlib.xla_extension' has no attribute 'DeviceArray'

**Comment by PhilipVinc** [comment]
It should not be a problem for most things.
However if you want to be sure, just install netket from here

```python
pip install -U 'git+https://github.com/netket/netket.git@compat'
```

I fixed it for you. Eventually it will end up in the official version but for a while it will be there

**Comment by JanaMensikova** [comment]
Ok, it seems it works. Thank you very much for your help!
